csharp/src/Drivers/BigQuery/BigQueryStatement.cs

```diff
@@ -0,0 +1,422 @@
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text.Json;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+using Apache.Arrow.Ipc;
+using Apache.Arrow.Types;
+using Google.Apis.Auth.OAuth2;
+using Google.Cloud.BigQuery.Storage.V1;
+using Google.Cloud.BigQuery.V2;
+using TableFieldSchema = Google.Apis.Bigquery.v2.Data.TableFieldSchema;
+using TableSchema = Google.Apis.Bigquery.v2.Data.TableSchema;
+
+namespace Apache.Arrow.Adbc.Drivers.BigQuery
+{
+    /// <summary>
+    /// BigQuery-specific implementation of <see cref="AdbcStatement"/>
+    /// </summary>
+    public class BigQueryStatement : AdbcStatement
+    {
+        readonly BigQueryClient client;
+        readonly GoogleCredential credential;
+
+        public BigQueryStatement(BigQueryClient client, GoogleCredential credential)
+        {
+            this.client = client;
+            this.credential = credential;
+        }
+
+        public IReadOnlyDictionary<string, string>? Options { get; set; }
+
+        public override QueryResult ExecuteQuery()
+        {
+            QueryOptions? queryOptions = ValidateOptions();
+            BigQueryJob job = this.client.CreateQueryJob(SqlQuery, null, queryOptions);
+            BigQueryResults results = job.GetQueryResults();
+
+            BigQueryReadClientBuilder readClientBuilder = new BigQueryReadClientBuilder();
+            readClientBuilder.Credential = this.credential;
+            BigQueryReadClient readClient = readClientBuilder.Build();
+
+            // TODO: translate the schema
+
+            string table = $"projects/{results.TableReference.ProjectId}/datasets/{results.TableReference.DatasetId}/tables/{results.TableReference.TableId}";
+
+            ReadSession rs = new ReadSession { Table = table, DataFormat = DataFormat.Arrow };
+            ReadSession rrs = readClient.CreateReadSession("projects/" + results.TableReference.ProjectId, rs, 1);
+
+            long totalRows = results.TotalRows == null ? -1L : (long)results.TotalRows.Value;
+            IArrowArrayStream stream = new MultiArrowReader(TranslateSchema(results.Schema), rrs.Streams.Select(s => ReadChunk(readClient, s.Name)));
+
+            return new QueryResult(totalRows, stream);
+        }
+
+        public override UpdateResult ExecuteUpdate()
+        {
+            BigQueryResults result = this.client.ExecuteQuery(SqlQuery, parameters: null);
+            long updatedRows = result.NumDmlAffectedRows == null ? -1L : result.NumDmlAffectedRows.Value;
+
+            return new UpdateResult(updatedRows);
+        }
+
+        public override object GetValue(IArrowArray arrowArray, Field field, int index)
+        {
+            if (arrowArray is Int64Array)
+            {
+                return ((Int64Array)arrowArray).Values[index];
```

Comment:
This will return the wrong value when the value is null. Use GetValue for Int64Array and DoubleArray.

Reply:
fixed in latest push